maildump/smtp.py

```diff
@@ -9,10 +10,91 @@
 log = Logger(__name__)
 
 
+class SMTPChannel(smtpd.SMTPChannel, object):
+    def __init__(self, server, conn, addr, smtp_auth, smtp_username, smtp_password):
+        super(SMTPChannel, self).__init__(server, conn, addr)
+        self._smtp_auth = smtp_auth
+        self._smtp_username = smtp_username
+        self._smtp_password = smtp_password
+        self._authorized = False
+
+    def smtp_EHLO(self, arg):
+        if not arg:
+            self.push('501 Syntax: EHLO hostname')
+            return
+        if self.__greeting:
+            self.push('503 Duplicate HELO/EHLO')
+            return
+        self.__greeting = arg
+        self.push('250-%s' % self.__fqdn)
+        if self._smtp_auth:
+            self.push('250-AUTH PLAIN')
+        self.push('250 HELP')
+
+    def smtp_AUTH(self, arg):
+        print >> smtpd.DEBUGSTREAM, '===> AUTH', arg
+        if not self._smtp_auth:
+            self.push('501 Syntax: AUTH not enabled')
+            return
+
+        if not arg:
+            self.push('501 Syntax: AUTH TYPE base64(username:password)')
+            return
+
+        if not arg.lower().startswith('plain '):
+            self.push('501 Syntax: only PLAIN auth possible')
+            return
+
+        auth_type, auth_data = arg.split(None, 1)
+        try:
+            auth_data = base64.b64decode(auth_data.strip())
+        except TypeError:
+            self.push('535 5.7.8 Authentication credentials invalid')
+            return
+
+        auth_data = auth_data.split('\x00')
+        if len(auth_data) == 3 and auth_data[0] == auth_data[1] and \
```

Comment:
I prefer the `if (...)` notation for wrapping instead of backslashes

Reply:
what's in the 0/1 fields of `auth_data`?

Reply:
hash given to AUTH PLAIN is:
`base64("username\x00username\x00password")`

So, `auth_data[0]` and `auth_data[1]` should be always username.